### ISSUE TITLE
.NET 6 release broke switch expression formatting

### ISSUE THREAD

**Issue by maloo** [dotnet/roslyn#57736]
Code like
```csharp
var x = new { p: "data" };
var y = x switch {
    { p: "123" } => 1,
    { p: "data"} => 2,
    _ => 3,
};
```
all end up on the same line after latest release, like this:
```csharp
var x = new { p: "data" };
var y = x switch { { p: "123" } => 1, { p: "data"} => 2,
    _ => 3,
};
```
Since this looks really bad I assume it is a bug. But if it is a breaking change, what setting should we update to get the former style?
We use .editorconfig from dotnet runtime plus the change: `csharp_new_line_before_open_brace=methods, types`. Do we need to add some new parameter here for switch expressions or patterns? We could not find any setting that made sense for switch/pattern.

**Comment by maloo** [comment]
> Moving to Roslyn since dotnet-format runs the Roslyn formatter.

Where can I find the Roslyn issue? Any reference?

**Comment by CyrusNajmabadi** [comment]
@maloo this is the issue.

**Comment by sharwell** [comment]
Support for this **.editorconfig** option was added in #53110. You'll need to add `object_collection_array_initializers` to the list assigned to `csharp_new_line_before_open_brace` (this value covers all expressions and patterns).

**Comment by maloo** [comment]
@sharwell But a switch expression is not an object initializer array. Adding this flag might fix this code, but it will add a newline in actual object initializer arrays which we explicitly didn't want.
There seems to be a flag missing for switch expressions. So I guess that should be added or code that is not object initializer should not be touched or generate check errors. I assume the ability to use editor config is that you should be able to use a code style different than the one in runtime.
And this worked fine in .NET 5 with dotnet format (--check)

**Comment by sharwell** [comment]
> But a switch expression is not an object initializer array

This is true. We would like to change the setting value to `expressions`, but there are compatibility concerns and it's an extremely rarely used setting.

**Comment by maloo** [comment]
But you already made a breaking change between .NET 5 and 6. dotnet format (used in CI by many) can no longer be used and there is no other linting option for dotnet.
1) Why do dotnet format (roslyn?) mess with expressions formats when there is no editor config setting for it, why not just leave the code as is until a setting is available.
2) A shared setting called expressions solves nothing, since it basically force everyone to use { on a new line for all types of expressions. I know dotnet/runtime prefers { on a newline. But forcing everyone to do that (with a breaking change), then why have the option to control this setting at all? When you put { on the same line as previous you do that for object initializer too. But that is very different from putting every switch expression on one line just because it happens to use an object pattern matching.

**Comment by sharwell** [comment]
> Why do dotnet format (roslyn?) mess with expressions formats when there is no editor config setting for it, why not just leave the code as is until a setting is available.

`object_collection_array_initializers` is the setting for expressions. See #39557.

> A shared setting called expressions solves nothing, since it basically force everyone to use { on a new line for all types of expressions. I know dotnet/runtime prefers { on a newline. But forcing everyone to do that (with a breaking change), then why have the option to control this setting at all?

This is a difficult space, because there are a lot of other cases where it's likely the user does want the brace kept on the preceding line. We may add a forced exclusion in the formatter to not move the brace to the preceding line if:

1. The `{` is the first token of a switch arm
2. The containing `switch` statement or expression spans multiple lines

**Comment by maloo** [comment]
I would love to see a forces exclusion for these cases, since it would bring it back to compatibility with .NET 5. In the long run I would like more granular settings for expressions. As seen here, there are many types of expressions... and the current setting for object initializers is a good thing, don't change that. Instead add new ones as needed to replace forced exclusions. But I think the forced exclusion would allow people that use { on previous line to use dotnet format and use switch expressions.

**Comment by sharwell** [comment]
> In the long run I would like more granular settings for expressions.

We've had a few requests for this over time. Eventually things settled into two groups:

1. People who wish to see granular settings expanded to new options, since it's a natural extension of how things exist today
2. People who have attempted to implement and test a granular settings expansion in the formatter

We find the group of people who enter (2) give up when they see how overwhelmingly complex this is and become heavily opposed to any attempt to make new formatter options.
